x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/PruneColumns.java

```diff
@@ -148,24 +96,143 @@ public LogicalPlan apply(LogicalPlan plan) {
+                    NamedExpression firstAggregate = aggregate.aggregates().getFirst();
+                    remaining = List.of(new Alias(firstAggregate.source(), firstAggregate.name(), attribute, firstAggregate.id()));
+                    p = aggregate.with(aggregate.groupings(), remaining);
+                }
+            } else {
+                if (inlineJoin && aggregate.groupings().containsAll(remaining)) { // not expecting high groups cardinality
+                    // It's an INLINEJOIN and all remaining attributes are groupings, which are already part of the IJ output (from the
+                    // left-hand side).
+                    if (aggregate.child() instanceof StubRelation stub) {
+                        var message = "Aggregate groups references ["
+                            + remaining
+                            + "] not in child's (StubRelation) output: ["
+                            + stub.outputSet()
+                            + "]";
+                        assert stub.outputSet().containsAll(Expressions.asAttributes(remaining)) : message;
+
+                        p = emptyLocalRelation(aggregate);
+                    } else {
+                        // There are no aggregates to compute, just output the groupings; these are already in the IJ output, so only
+                        // restrict the output to what remained.
+                        p = new Project(aggregate.source(), aggregate.child(), remaining);
+                    }
+                } else { // not an INLINEJOIN or there are actually aggregates to compute
+                    p = aggregate.with(aggregate.groupings(), remaining);
+                }
+            }
+        }
+
+        return p;
+    }
+
+    private static LogicalPlan pruneColumns(InlineJoin ij, AttributeSet.Builder used, Holder<Boolean> recheck) {
+        LogicalPlan p = ij;
+
+        used.addAll(ij.references());
+        var right = pruneColumns(ij.right(), used, true);
+        if (right.output().isEmpty()) {
+            p = ij.left();
+            recheck.set(true);
+        } else if (right != ij.right()) {
+            // if the right side has been updated, replace it
+            p = ij.replaceRight(right);
+        }
+
+        return p;
+    }
+
+    private static LogicalPlan pruneColumns(Eval eval, AttributeSet.Builder used, Holder<Boolean> recheck) {
+        LogicalPlan p = eval;
+
+        var remaining = pruneUnusedAndAddReferences(eval.fields(), used);
+        // no fields, no eval
+        if (remaining != null) {
+            if (remaining.isEmpty()) {
+                p = eval.child();
+                recheck.set(true);
+            } else {
+                p = new Eval(eval.source(), eval.child(), remaining);
+            }
+        }
+
+        return p;
+    }
+
+    private static LogicalPlan pruneColumns(Project project, AttributeSet.Builder used) {
```

Comment:
I think this is the meat of the PR, right?

Reply:
Handling `Project` is part of it, but the main change is calling (the now only method called) `pruneColumns()` within the (now renamed) `pruneColumnsInInlineJoin()` and using updated `used` set, since `INLINEJOIN` can be followed by any other command and each of these will need to have columns inspected for potential pruning, _but_ aware of the fact that the pruning is done following INLINEJOIN; this "awareness" allows dropping the INLINEJOIN node when its output is entirely contained within the left-hand side of the join already.